Remove componentWillReceiveProps from ColorPicker

oandregal opened the pull request:
We want to get rid of `componentWillReceiveProps` method https://github.com/WordPress/gutenberg/issues/11360

### Some context

This is the `ColorPicker` tree:

`ColorPicker`
* `Saturation`
* `Hue`
* `Inputs`
  * `Input` (for HEX)
  * `Input` (for RGB - actually, three inputs, one for each band)
  * `Input` (for HSL - actually, three inputs, one for each band)

The `ColorPicker` acts as a coordinator of the different color views (`Saturation`, `Inputs`) and keeps them in sync - when one changes the color, the others are updated. It uses some utils to work with colors (such as converting `#fff` to `#ffffff` and so on).

Besides those updates, the `Input` leaf components were updated when some internal state changed. They used this internal state as a cache. They didn't notify the `ColorPicker` of changes on every keystroke, only for certain kind of events (when the users hit the `ENTER` key, on blur, etc).

### The problem

The `Input` leaf component was updated both on state and prop changes. It took advantage of `componentWillReceiveProps` only executing when props changed to differentiate state and props changes. This is considered an [anti-pattern](https://reactjs.org/blog/2018/06/07/you-probably-dont-need-derived-state.html?no-cache=1#anti-pattern-erasing-state-when-props-change) in React and the reason why they deprecated `componentWillReceiveProps` in favor of `getDerivedStateFromProps`. 

Unlike `componentWillReceiveProps`, `getDerivedStateFromProps` is called when either props **or** state changes, so we no longer can distinguish why the method is called (is it a prop change or a state change?) within the same component. To deal with this use case React recommends centralizing all data in one place, either in a component state (they call this _uncontrolled component_) or props (they call this _controlled component_).

### How this PR fixes it

We wanted the `Input` leaf components to react to changes in other views (`Saturation`, for example) by updating their props. To achieve this I needed to do two things:

- Converting the `Input` components into controlled components. Lift up the cache (internal state) to the `ColorPicker` by distinguishing when a change needs to be `commited` (spread to the other views) or when a change is a `draft` (should only affect the `Inputs` view).
- Centralizing all the logic to decide when to commit changes to the colors. It was previously spread through several components, for example, the `Inputs` component checked [if the Hex code was valid](https://github.com/WordPress/gutenberg/pull/11772/files#diff-2a6da6e04df0dda01acf1714b4d7d4b0L122) and the `ColorPicker` checked whether the RGB/HSL was valid to back out from notifying all the views of the color change.

### Testing

* Test that changes in one view update the others.
* Test color changes only the keyboard, to make sure focus is preserved.

jorgefilipecosta commented on packages/components/src/color-picker/inputs.js:
```diff
@@ -18,24 +18,18 @@ import IconButton from '../icon-button';
 		this.handleKeyDown = this.handleKeyDown.bind( this );
 	}
 
-	componentWillReceiveProps( nextProps ) {
```
I think removing componentWillReceiveProps,  has a small behavior change. Previously if the value was changed externally after the constructor is executed ( e.g: a plugin changing the block attribute value), to contain uppercase letters we would make the letters lowercase. Now the letters continue uppercase.
I think to solve this we would need getDerivedState from props. I'm not certain if it is worth it as it would have a performance impact and this approach seems fine.

gziolo replied:
I have no idea if I'm right anymore :) This logic is quite complex and hard to follow. There are some cases where you trigger change outside and try to mirror local state and in other places it only updates local state 🤷‍♂️

nosolosw replied:
Just to note that I'm investigating this and trying to come up with some tests that guarantee behavior works as expected.

nosolosw replied:
Tests added so it's now clear what fails on removing `componentWillReceiveProps` method and what happens when it's substituted by `getDerivedStateFromProps`.

Will ping you for a review when all tests pass. I hadn't done that yet because I was still working on it. Marking it as "In Progress" for clarification.

jorgefilipecosta reviewed (dismissed):
Things look good in my tests 👍

oandregal committed:
Tests for onChange prop

oandregal committed:
Tests for onChange prop not being called

oandregal committed:
Add is rendered tests

oandregal committed:
Input values are lowercased on mount

oandregal committed:
Input values are lowercased on subsequent props updates

oandregal committed:
Remove unnecessary onChange declaration

oandregal committed:
Extract toLowerCase function

oandregal committed:
Remove componentWillReceiveProps lifecycle method

oandregal committed:
Create a separate path for handling input data

oandregal committed:
Add a state prop to the data lifted up

Originally only uses 'commit' but a 'draft' state
will be introduced in next steps.

oandregal committed:
Add draft values to state

oandregal committed:
Make Input a fully controlled component

Lifts the Input internal state up to the ColorPicker.

oandregal committed:
Flatten draft values

otherwise setState( { draftValues: { rgb } } ) makes
draftValues.hsl and draftValues.hex undefined.

oandregal committed:
Consider handleChange event always a draft operation

oandregal committed:
Lift HEX validation up

We need all logic to be in one place.

nosolosw commented:
Ugh, this has taken more time than I expected. The `CodePicker` component tree suffered from a couple of smells that made it difficult to work with. I've updated the description to better explain the changes here so you may want to read that first.

I still need to update the tests I've added in this PR to the new changes, but wanted to give you as much time as possible for review.

nosolosw commented on packages/components/src/color-picker/index.js:
```diff
@@ -43,30 +43,81 @@ import Alpha from './alpha';
+	handleInputChange( data ) {
+		if ( data.state === 'commit' ) {
+			this.handleChange( data );
+		} else if ( data.state === 'draft' ) {
```
By lifting the draft cache to the `ColorPicker` and updating state we are forcing the other views to re-render, which may cause performance issues. We should test and avoid re-renderings if necessary.

notnownikki replied:
Yes, we're re-rendering saturation and hue every time an input changes. Can we keep the draft in the Inputs state until we're ready to commit, and then call this.props.onChange?

It seems to me that `ColorPicker` knows too much about the child components. If the child components only took a hex value color, they could process it into whatever values they needed. For example, `Inputs` could split it into rgba, and when the internal draft state was committed by calling `this.props.onChange`, the `ColorPicker` state would change, which would send the new `color` to all the child components, and they would update their internal state with `getDerivedStateFromProps`.

nosolosw replied:
I wish we could. This PR was so involved because we cannot longer do that. I've described why we can't do that in this issue description. Does that help?

In 520c3e1 and be0cc4f I've prevented the other components from re-render, which alleviates my original concerns.

notnownikki replied:
Good that the re-render is fixed :)

I'm sorry, I totally misunderstood the description. You're totally right. Wow, that seems like quite a step backwards, to me. Ok, I understand now, thanks :)

oandregal committed:
Test ColorPicker: use uppercase HEX color value

to make sure generated output lowercases the input color.

oandregal committed:
Test ColorPicker: uses React TestRenderer instead of ShallowRenderer

oandregal committed:
Test ColorPicker: draft changes onChange event

oandregal committed:
Test ColorPicker: commit changes onBlur

oandregal committed:
Test ColorPicker: commit changes onKeyDown UP

oandregal committed:
Test ColorPicker: commit changes onKeyDown DOWN

oandregal committed:
Test ColorPicker: commit changes onKeyDown ENTER

oandregal committed:
Test Input: update tests

nosolosw commented on packages/components/src/color-picker/inputs.js:
```diff
@@ -15,54 +15,46 @@ import { DOWN, ENTER, UP } from '@wordpress/keycodes';
 	handleChange( value ) {
 		const { valueKey, onChange } = this.props;
-		// Protect against expanding a value while we're typing.
-		if ( value.length > 4 ) {
```
Previous to this removal, the component will tell `ColorPicker` to update the other views (and the `onChangeComplete` callback) when the HEX value introduced was longer than 4 characters. I'm looking into port this behavior.

nosolosw replied:
Fixed in d06c6f9.

oandregal committed:
Fire commit or draft change depending on the value length

oandregal committed:
Delete unused snapshots tests

oandregal committed:
Test Input: update tests for different value lengths

oandregal committed:
Make Saturation, Hue, and Alpha pure components

so they don't get re-rendered when draft values change.

oandregal committed:
Make Icon pure

oandregal committed:
Update snapshot

nosolosw commented:
I've fixed all the concerns I had about this PR and this is ready for review/merge.

jorgefilipecosta reviewed (commented):
Hi @nosolosw,
It looks like there is a regression in these changes:
Go to the color picker in hex mode.
Write a random string e.g: dwasaasas press enter.
Try to delete the string using backspace and verify it is impossible to delete the last character.
![jan-21-2019 10-18-02](https://user-images.githubusercontent.com/11271197/51468228-5d423300-1d66-11e9-86f0-48c2b435fc9a.gif)

nosolosw commented:
@jorgefilipecosta I'll revisit this by the end of the week or first thing next week. It's been a few months since I worked on it so I'll need some time to refresh how this all worked. I'm in the middle of something else that I need to finish first, if that's ok.

oandregal committed:
Allow introducing empty values

By splitting handleChange in smaller methods, we can target
which value to update without relying on the data properties.
This brings back the ability to add empty values.

oandregal committed:
Do not commit empty values

oandregal committed:
Reset draft values when views are toggled.

This prevents leaving views with invalid or empty values.

oandregal committed:
Consider Saturation and Hue when the source=rgb

nosolosw commented:
@jorgefilipecosta addressed your feedback, this is ready for review.

jorgefilipecosta commented on packages/components/src/color-picker/inputs.js:
```diff
@@ -97,63 +107,115 @@ export class Inputs extends Component {
+		} );
+	}
+
+	handleRChange( data ) {
```
Can we use a higher-order function to auto-generate these functions ( handleRGB/HSL changes ) and avoid duplicate code?
e.g:
```
handleChange( source, component ) {
	return function ( data ) {
		this.props.onChange( {
			...this.props[source],
			sourc,
			state: data.state,
			[component]: data[component]
		} );
	}
}
handleGChange = handleChange( 'rgb', 'g');
```

nosolosw replied:
I thought about that and it's a smart use of object destructuring. Thanks for sharing! My concern is that it compresses the code to the point it becomes difficult to reason about: it's difficult to see the object shape for each handler and that `data.r` will override `props.rgb.r` for the R handler in RGB. I lean towards optimizing code for readers. Is this a blocker for you?

nosolosw replied:
Also: we can't use that for `handleAChange`.

mcsf replied:
```js
handleChange( source, key, data ) {
	const value = ( NORMALIZERS[ key ] || identity )( data[ key ] );
	this.props.onChange( {
		...this.props[ source ],
		[ key ]: value,
		source,
		state: data.state,
	} );
}
```

with

```js
const NORMALIZERS = {
	a: ( value ) =>
		Math.round( Math.max( 0, Math.min( 1, value ) ) * 100 ) / 100,
};
```

This could naturally be adapted to key not just on `key` but on `( source, key )`, if need be.

Then:

```js
onChange={ this.handleChange.bind( this, 'rgb', 'a' ) }
```

mcsf replied:
>  I lean towards optimizing code for readers. I don't feel strongly about this and I'm open to change. Is this a blocker for you?

I agree we need to optimise for readability. I think that feeling the need to duplicate and specialise in order to improve readability is a code smell, though. The generalisation—`handleChange` or whatever we use—needs to be clearer.

nosolosw replied:
@jorgefilipecosta @mcsf Thanks for the push to simplify. I've pushed a new version that hopefully offers a better code compression without compromising readability.

jorgefilipecosta reviewed (commented):
I'm facing a strange regression in my tests, I'm not sure yet what is the reason.
To reproduce.
Add a paragraph.
Open the background custom color picker.
Write 'red' in the hex input field press enter and verify the color changes to red.
Try to change color dragging the circle inside color picker and verify it is not possible.
![jan-30-2019 19-56-40](https://user-images.githubusercontent.com/11271197/52008748-3522a480-24c9-11e9-940d-629b224c3c13.gif)

nosolosw commented:
>  Try to change color dragging the circle inside color picker and verify it is not possible.

@jorgefilipecosta I tried with Firefox and Chrome and can't reproduce. That was an issue in earlier commits but I've fixed it before pinging you for review. Could you perhaps try nuking your local branch and start fresh? Also refreshing browser cache?

oandregal committed:
Rename handleChange to commitValues

oandregal committed:
Extract set and reset draft values

oandregal committed:
Use switch instead of if/else

oandregal committed:
Refactor source to be taken from view

oandregal committed:
Refactor how the colors object is updated.

Before this change, the input components were responsible to send
the value that changed and the untouched ones. After this change,
the input components only send the value that changed
and the index is responsible to merge this into the old color object.

oandregal committed:
Fix linter issue: dependencies comment block

oandregal committed:
Update tests to new API

mcsf commented:
Thanks for circling back to this, @nosolosw. It has improved a lot.

That said, I'm still worried about the changes, but now I realise that the complexity in this PR stems from pre-existing complexity around `ColorPicker` since its inception (#10564).

Some of the flow is currently a little more explicit, with the idea of drafts and committing changes, and with the centralisation of state. However, I am concerned with the ratio of derived vs. essential state. A couple of questions (for which you personally don't have to have an answer):

- Is the concept of drafting-and-committing related to a feature requirement? That is, is there an expected feature whereby a user can compose a colour by varying its components without the previewed colour changing, until the user _confirms_ these new values and the new colour is previewed?

I ask this because in `master` I get an inconsistent experience: when varying the values of colour components with the keyboard, both `ColorPicker` and the affected block vary in real time. However, when I use the mouse (in Safari, at least) on the input's [native Increment/Decrement buttons](https://user-images.githubusercontent.com/150562/53252926-f80c9500-36b7-11e9-86a0-fc058387ba29.png), I get no colour response whatsoever.

- Assuming no such feature is expected, and thus that changing the colour inputs should update the picker and block in real time, is there any other reason to duplicate colour state? There are several ways in which we collect derived state:

1. By holding state for one same colour in different colour models (RGB and its hex form, HSL and HSV).
2. By holding these data in conjunction with the `color` datum.
3. By caring about the `source` from which a colour was set (?).
4. By holding all this in conjunction with draft data for each colour representation (`draftHex`, `draftRgb`, `draftHsl`).

The overall picture of `ColorPicker`'s data model is given by its state [initialisation](https://github.com/WordPress/gutenberg/blob/8462f88500dc31ad603b81c5883d2dedf7a7d115/packages/components/src/color-picker/index.js#L105-L110) and the [`colorToState`](https://github.com/WordPress/gutenberg/blob/8462f88500dc31ad603b81c5883d2dedf7a7d115/packages/components/src/color-picker/utils.js#L59-L67) util. State duplication requires a lot of manual synchronisation and impacts the overall maintainability of the component.

***

I wanted to provide a more informed review, and so I circled back to a jsfiddle sandbox. In it, I roughly mimicked the composition of the `ColorPicker` component into: a colour preview (analogous to `Saturation`); a slider for the alpha channel; a slider for the hue; and an editable field set for whichever colour model is active. I believe it answers all the bigger questions in terms of state centralisation, controlling child components, and handling different colour models agnostically. 

Crucially, only one component is stateful: `ColorPicker` itself. Its state has the following shape: `{ mode: 'hex',
values: ['#ffff00'], alpha: 100 }`.
- The alpha channel is thus treated separately from the chosen colour model.
- `mode` can be `'hex'`, `'rgb'`, `'hsl'`.
- `values` is always an array and always corresponds to `mode`. If `mode` were `'hsl'` then `values` could be `[60, 100, 50]`.
- `ColorPicker` has no knowledge specific to any colour models, and neither does `PickerForm`; instead, an external config dictates how any colour model relates to basic hexadecimal.

Here is the sandbox in question: https://jsfiddle.net/mcsf/qa67mtwL/

With that, I'd like to help move this forward and not just feel like a blocker 😄, so:
- @nosolosw, does it feel like what I shared should change the direction of this PR? If so, in what way?
- I can help with guidance and code. Do you need any?
- @ryelle, I'm looping you in because you may have more context for the questions I raised earlier in this comment.

nosolosw commented:
@mcsf On `master`, the input components (`hex`, the single `rgb` and `hsl` components) do differentiate between _draft_ and _commit_ states - the code is just not that explicit as it is in this PR. When interacting with the keyboard, they're only updated when: 1) the `INTRO`, `UP` or `DOWN` keys are pressed, 2) they lost focus, or 3) value introduced is longer than 4 characters.

This PR was created to get rid of `componentWillReceiveProps`, so it intends to be a pure refactor.

Given this new information, would your review change?

mcsf reviewed (commented):
> On `master`, the input components […] do differentiate between _draft_ and _commit_ states - the code is just not that explicit

Cool, that was my impression.

> When interacting with the keyboard, they're only updated when: 1) the `INTRO`, `UP` or `DOWN` keys are pressed, 2) they lost focus, or 3) value introduced is longer than 4 characters.

Thanks for detailing. What I'd like to know for sure is which of these characteristics are first-class requirements vs. just consequences of implementing other requirements.

> This PR was created to get rid of `componentWillReceiveProps`, so it intends to be a pure refactor.

Yeah, I'm aware, but it felt like we should stop and ask these questions, because potentially we could achieve that same goal while also simplifying things.

> Given this new information, would your review change?

I can prioritise a review of the refactor itself, if you'd like this off your plate. Since there a bunch of moving parts, having a second person to also QA this would be great.

nosolosw commented:
> Thanks for detailing. What I'd like to know for sure is which of these characteristics are first-class requirements vs. just consequences of implementing other requirements.

That's great and I appreciate it. The `ColorPicker` hierarchy does have some gotchas/bugs that could be polished. I hope this refactor brings _some_ clarity over the existing code and that it helps to address those issues in subsequent PRs. This component's interactions are quite complex. I can work on improving them in a following PR if that's something that should be prioritized! I just don't think we should tackle that big a change in a single step.

nosolosw commented on packages/components/src/color-picker/index.js:
```diff
@@ -43,30 +43,144 @@ import Alpha from './alpha';
+
+		if ( isValueEmpty( data ) ) {
+			return;
+		}
```
No, it's because of the former `handleChange` method in the inputs component, see https://github.com/WordPress/gutenberg/pull/11772/files#diff-2a6da6e04df0dda01acf1714b4d7d4b0L120 Note how the if statements checked for non-empty values as in `if ( data.hex )`.

The former HEX case you mention was also lifted up and is now part of the `isValidColor` check.

mcsf replied:
Sounds good.

nosolosw commented:
Rebased from `master` as there was a fix for the unrelated failing e2e test.

jorgefilipecosta reviewed (commented):
Hi @nosolosw, I think I understood the logic behind these changes, and it seems correct. I think the component become complex over time and there is an opportunity to simplify the component based on the code the fiddle that @mcsf shared.
But for now, maybe we can unblock this deprecated react API usage.
In my tests, I just found a problem but it may be something on my machine, I can not change colors #000000 and #ffffff using on the color picker.
I pasted the following block on the code editor:
```
<!-- wp:paragraph {"customTextColor":"#ffffff","customBackgroundColor":"#000000"} -->
<p style="background-color:#000000;color:#ffffff" class="has-text-color has-background">aaaaaaa</p>
<!-- /wp:paragraph -->
```
I oppened the background color picker and the text color picker. I checked that by using the mouse in the color area I was not able to change the colors:
![mar-01-2019 19-43-23](https://user-images.githubusercontent.com/11271197/53662984-4f80a700-3c5c-11e9-95ec-2528ae871431.gif)


Coincidence or not the fiddle share by @mcsf also seems affected by this problem:
![mar-01-2019 19-58-10](https://user-images.githubusercontent.com/11271197/53663020-6f17cf80-3c5c-11e9-8b38-7b8e2bd9535d.gif)

I reproduced the problem in chrome and safari.
On master, things work great. 

@mcsf, @nosolosw are you able to reproduce this problem? Or it is something strange happening in my test environment?

mcsf commented:
> In my tests, I just found a problem but it may be something on my machine, I can not change colors #000000 and #ffffff using on the color picker.

I can reproduce this too, good call.

> Coincidence or not the fiddle share by @mcsf also seems affected by this problem:

It's a partial coincidence. It stems from the fact that, when our current colour is white, black, or any grey, in HSL this amounts to `hsl(x, y, 100)`, `hsl(x, y, 0)`, `hsl(x, 0, z)`, respectively. Notably, the hue `x` can have any value without affecting the actual resulting colour. Since the fiddle uses `#123456` as the base format, changes to the hue slider when dealing with greys have no effect on that hexadecimal state, and thus the UI can never react with a new hue value.

As far as I could tell when I last looked at Gutenberg's `ColorPicker`, this issue affects it differently. It tracks `oldHue` (presumably for this sort of edge case?), as well as different models for the same colour. So, at least without looking at the diff again, I can't pinpoint what caused this regression.

For what it's worth, I've updated my [fiddle](https://jsfiddle.net/mcsf/qa67mtwL/) to handle this issue in a different way, without expanding its state. Instead, I've added a rule so that, whenever the hue slider is used, the saturation and lightness must conform to certain bounds:

```js
function onHueChange(event) {
  const hue = parseInt(event.target.value, 10)
  const newHslValues = [
    hue,
    constrict(10, 100, hslValues[1]), // Avoid greys
    constrict(10, 90, hslValues[2]), // Avoid black and white
  ]
  let values = convert('hsl', mode, newHslValues)
  setValues(values)
}
```

In the case of the hue slider, I think this is a fair requirement. Aside from not "compromising" the code, it automatically "pushes" a colour out of the greys and thus the user can actually see a slight hue changing as the slider moves.

![color-picker-constrict](https://user-images.githubusercontent.com/150562/53817999-48f57680-3f5f-11e9-8dcc-c1a0d24f7330.gif)

The reason I'm showing the above is that this could inform subsequent refactors to `ColorPicker`. For now, obviously, the priority is to fix the present PR so that the Saturation matrix doesn't lock up.

ryelle commented on packages/components/src/color-picker/index.js:
```diff
@@ -43,30 +43,144 @@ import Alpha from './alpha';
+	( data.source === 'rgb' && (
+		( ! data.r || ! data.g || ! data.b ) &&
+		( ! data.h || ! data.s || ! data.v || ! data.a ) &&
+		( ! data.h || ! data.s || ! data.l || ! data.a )
```
These values above can be zero, so this is causing some trouble with using the keyboard to move through the saturation. If you focus into the saturation handle, you can use <kbd>shift</kbd> + <kbd>left</kbd> or <kbd>shift</kbd> + <kbd>right</kbd> (or just left/right) to move the handle, but you can't get it all the way to the edge.

![colors-key](https://user-images.githubusercontent.com/541093/53823127-1fc4ee00-3f3f-11e9-9821-715b050c6856.gif)

I'm pretty sure this is also [the root of this issue,](https://github.com/WordPress/gutenberg/pull/11772#pullrequestreview-209761283) both seemed fixed if I change this to check for [`isFinite`](https://lodash.com/docs/4.17.11#isFinite)

oandregal committed:
Remove unnecessary check for Saturation and Hue

nosolosw commented:
@jorgefilipecosta that was a good catch, thanks! :)

@jorgefilipecosta @mcsf @ryelle It took me a few days to get back to this, sorry about that. The bug was caused by an extra check (`isEmptyColor`) for Saturation and Hue that I had inadvertently added in the refactor and that isn't present in `master`. I tested this on FF and Chrome (Ubuntu Linux):

![Peek 2019-03-15 12-48](https://user-images.githubusercontent.com/583546/54429644-67a6fa80-4721-11e9-82c8-e651dbd81e01.gif)

![Peek 2019-03-15 12-49](https://user-images.githubusercontent.com/583546/54429648-6b3a8180-4721-11e9-80f2-9f4e529afe76.gif)

nosolosw commented:
@jorgefilipecosta this is ready for a review.

jorgefilipecosta reviewed (approved):
In my tests, I did not found any regression so I think this PR is ready. Thank you @nosolosw for all the iterations performed. This is a nice step in removing deprecated react methods.

As noted during the discussions, we may have an opportunity to simplify the way these components work based on the prototype code shared.
